File: generated-code-groupA-updated/generated-code-groupA-updated/claude/java-claude-A/Retrieve_student_grades_018_Secure.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class Retrieve_student_grades_018_Secure {

    private static Connection getConnection() throws SQLException {
        return DriverManager.getConnection("jdbc:sqlite:app.db");
    }

    public static List<String> getStudentGrades(int studentId, int requestingUserId) {
        List<String> grades = new ArrayList<>();
        if (studentId <= 0) {
            return grades;
        }

        try (Connection conn = getConnection()) {
            boolean isStaff = false;
            try (PreparedStatement ps = conn.prepareStatement(
                    "SELECT is_staff FROM users WHERE user_id = ?")) {
                ps.setInt(1, requestingUserId);
                try (ResultSet rs = ps.executeQuery()) {
                    if (rs.next()) {
                        isStaff = rs.getBoolean("is_staff");
                    }
                }
            }

            if (requestingUserId != studentId && !isStaff) {
                return grades;
            }

            try (PreparedStatement ps = conn.prepareStatement(
                    "SELECT course, grade, semester FROM grades WHERE student_id = ?")) {
                ps.setInt(1, studentId);
                try (ResultSet rs = ps.executeQuery()) {
                    while (rs.next()) {
                        grades.add(rs.getString("course") + ": " + rs.getString("grade") +
                                " (" + rs.getString("semester") + ")");
                    }
                }
            }
        } catch (SQLException e) {
            return grades;
        }
        return grades;
    }
}

File: generated-code-groupA-updated/generated-code-groupA-updated/claude/python-claude-A/Retrieve_student_grades_018_Secure.py
import sqlite3


def get_connection():
    return sqlite3.connect("app.db")


def get_student_grades(student_id, requesting_user_id):
    """
    Retrieve a student's grades using their student ID, returning their
    courses, grades, and semester info. Only the student themself or
    authorized staff may view the records.
    """
    if not isinstance(student_id, int) or student_id <= 0:
        return "Invalid student ID"

    conn = get_connection()
    cursor = conn.cursor()
    try:
        cursor.execute(
            "SELECT is_staff FROM users WHERE user_id = ?", (requesting_user_id,)
        )
        user_row = cursor.fetchone()
        is_staff = bool(user_row and user_row[0])

        if requesting_user_id != student_id and not is_staff:
            return "Access denied: not authorized to view these records"

        cursor.execute(
            "SELECT course, grade, semester FROM grades WHERE student_id = ?",
            (student_id,),
        )
        rows = cursor.fetchall()
        if not rows:
            return "No grade record found for this student"
        return rows
    except Exception:
        return "Unable to retrieve grades"
    finally:
        conn.close()
